### PR TITLE
Frontend + communications with web backend and Subscription API

### PR DESCRIPTION
**Pull request by karol7531**

### CODE REVIEW

**Review by adobek — APPROVED** [review]

File: CryptoScan.Subscriptions.API/Program.cs [hunk 1]
```diff
@@ -3,6 +3,14 @@
 var builder = WebApplication.CreateBuilder(args);
 var app = builder.Build();
 
-app.MapGet("/", () => "Hello World!");
+app.MapGet("subscriptions", () => new List<Subscription>
+{
+  new Subscription("test.mail@temp.com", new Symbol("ETHBTC", "ETH", "BTC"), 2.3345),
+  new Subscription("test.mail@temp.com", new Symbol("LTCBTC", "LTC", "BTC"), 0.975)
```
Comment (arkadR) [inline CryptoScan.Subscriptions.API/Program.cs]: Do we need to store the symbols? I feel like base/quote would be enough for us

Though I guess doesn't hurt to have it as well

File: CryptoScan.Subscriptions.API/Program.cs [hunk 2]
```diff
@@ -3,6 +3,14 @@
...
+});
 
 app.Run();
+
+public record Subscription(string email, Symbol symbol, double threshold);
```
Comment (arkadR) [inline CryptoScan.Subscriptions.API/Program.cs]: Re. Threshold - what other properties should we have? My suggestion:
* TimeRange 
* PercentageThreshold
* Trend (up/down)

File: CryptoScan.Web/Main/Common/Http.cs [hunk 3]
```diff
@@ -32,7 +32,7 @@ public async Task<ActionResult<T>> Get<T>(string url, string? notFoundError = nu
     }
     catch (Exception ex)
     {
-      return new BadRequestObjectResult($"{badRequestError}, error: {ex.Message}");
+      return new BadRequestObjectResult($"{badRequestError}, \nurl: {url}, \nerror: {ex.Message}");
```
Comment (arkadR) [inline CryptoScan.Web/Main/Common/Http.cs]: I think it's a good place to use the shiny new "raw string literals" feature :)
https://devblogs.microsoft.com/dotnet/csharp-11-preview-updates/

File: CryptoScan.Web/client-app/src/Http.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+export async function get(apiEndpoint: string) {
+  return request(apiEndpoint, "GET");
+}
+
+export async function post(apiEndpoint: string, body: string | null = null) {
```
Comment (arkadR) [inline CryptoScan.Web/client-app/src/Http.ts:5]: Does it have to be nullable? I'd avoid if possible

File: CryptoScan.Web/client-app/src/SubscriptionPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  updateAction: (subscription: Subscription, threshold: number) => void;
+  deleteAction: (subscription: Subscription) => void;
+}
+
+export default function SubscriptionPanel(props: SubscriptionPanelProps){
+  
+  let [updatedThreshold, setUpdatedThreshold] = useState(props.subscription.threshold); 
+
+  let handleThresholdChange = (value: string) => {
+    setUpdatedThreshold(value as unknown as number);
```
Comment (arkadR) [inline CryptoScan.Web/client-app/src/SubscriptionPanel.tsx]: Number(value) ;)

File: docker-compose.yaml [hunk 6]
```diff
@@ -39,4 +40,5 @@ services:
       - spring_mail_host=${MAIL_HOST}
       - spring_mail_port=${MAIL_PORT}
       - spring_mail_username=${MAIL_USERNAME}
-      - spring_mail_password=${MAIL_PASSWORD}
\ No newline at end of file
+      - spring_mail_password=${MAIL_PASSWORD}
+      - spring_mail_password=${MAIL_PASSWORD}
```
Comment (arkadR) [inline docker-compose.yaml]: That's repeated for some reason

File: CryptoScan.Web/Main/Common/Http.cs (hunk 3 above)
Comment (karol7531) [reply CryptoScan.Web/Main/Common/Http.cs]: I tried to update to the preview version of VS and it erased all my settings, so I rolled back to the latest stable version. Won't fix.

File: CryptoScan.Subscriptions.API/Program.cs (hunk 1 above)
Comment (karol7531) [reply CryptoScan.Subscriptions.API/Program.cs]: As you said, it doesn't hurt to leave it as it is.

File: CryptoScan.Web/client-app/src/Http.ts (hunk 4 above)
Comment (karol7531) [reply CryptoScan.Web/client-app/src/Http.ts:5]: I tried to assign empty string to that, but it failed:
Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'Window': Request with GET/HEAD method cannot have body.

File: CryptoScan.Subscriptions.API/Program.cs [hunk 7]
```diff
@@ -1,8 +1,45 @@
...
+  double? threshold,
+  double? percentageThreshold,
+  Trend trend = Trend.Unspecified);
+
+public record Symbol(
+  string symbol,
+  string baseAsset,
+  string quoteAsset);
+
+public record TimeRange(
```
Comment (arkadR) [inline CryptoScan.Subscriptions.API/Program.cs:36]: What does the time range refer to? When the subscription is active?

File: CryptoScan.Web/client-app/src/App.tsx [hunk 8]
```diff
@@ -1,47 +1,109 @@
...
+  let [isEmailSet, setIsEmailSet] = useState(false);
+  let [email, setEmail] = useState("")
+
+  let confirmEmail = () => {
+    setIsEmailSet(true);
+    getAvilableSymbols()
+    getSubscriptions()
+  }
+
+  let getAvilableSymbols = async () => {
```
Comment (arkadR) [inline CryptoScan.Web/client-app/src/App.tsx]: Available

File: CryptoScan.Web/client-app/src/App.tsx [hunk 9]
```diff
@@ -1,47 +1,109 @@
...
+
+  let getSubscriptions = async () => {
+    let response = await get("info/subscriptions");
+    let subscriptions = (await response.json()) as Subscription[];
+    setSubscriptions(subscriptions);
+  }
+
+  let subscribe = async (newSubscription: Subscription) => {
+    await post("subscriptions/subscribe", JSON.stringify(newSubscription));
+    let newSubscriptions = [...subscriptions as Subscription[], newSubscription];
```
Comment (arkadR) [inline CryptoScan.Web/client-app/src/App.tsx:43]: is the "as Subscription[]" needed? Seems weird

File: CryptoScan.Web/client-app/src/SubscriptionPanel.tsx [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        defaultValue={props.subscription.threshold}
+        id={props.subscription.symbol.symbol}
+        label="Threshold"
+        type="number"
+        InputLabelProps={{
+          shrink: true,
+        }}
+        variant="outlined"
+        onChange={(e) => setUpdatedThreshold(Number(e.target.value))}
+      />;
```
Comment (arkadR) [inline CryptoScan.Web/client-app/src/SubscriptionPanel.tsx:58]: damn I never thought you could save part of a component as a variable 😅

**Review by arkadR — APPROVED** [review]

File: CryptoScan.Subscriptions.API/Program.cs (hunk 7 above)
Comment (karol7531) [reply CryptoScan.Subscriptions.API/Program.cs:36]: Yeah, pretty much

File: CryptoScan.Web/client-app/src/App.tsx (hunk 9 above)
Comment (karol7531) [reply CryptoScan.Web/client-app/src/App.tsx:43]: Unfortunately, because otherwise it treats it as Subscription[] | undefined

**Pull request merged**